File: backend/app/schemas/group.py
"""Pydantic schemas for group endpoints."""

from __future__ import annotations

from datetime import datetime  # noqa: TC003
from typing import Literal

from pydantic import BaseModel, field_validator

# ── Enriched list schemas ─────────────────────────────────────────────────────


class RoundSummary(BaseModel):
    id: str
    round_number: int
    status: str
    book_title: str | None = None
    book_author: str | None = None
    book_cover_url: str | None = None
    book_page_count: int | None = None


class MyReadingProgress(BaseModel):
    current_page: int | None = None
    total_pages: int | None = None
    percentage: float


class LastMessagePreview(BaseModel):
    sender_display_name: str | None = None
    sender_avatar_url: str | None = None
    content_text: str | None = None
    content_type: str
    created_at: datetime


class MemberAvatar(BaseModel):
    user_id: str
    display_name: str | None = None
    avatar_url: str | None = None


# ── Validate / Join schemas ────────────────────────────────────────────────────


class GroupValidateResponse(BaseModel):
    valid: bool
    name: str | None = None
    photo_url: str | None = None
    member_count: int = 0

    model_config = {"from_attributes": True}


class GroupJoinRequest(BaseModel):
    invite_code: str

    @field_validator("invite_code")
    @classmethod
    def validate_code(cls, v: str) -> str:
        v = v.strip().upper()
        if len(v) != 8:
            raise ValueError("Codigo deve ter 8 caracteres.")
        return v


class GroupJoinResponse(BaseModel):
    message: str
    group_id: str


# ── CRUD schemas ──────────────────────────────────────────────────────────────


class MemberSummary(BaseModel):
    user_id: str
    username: str | None
    display_name: str | None
    avatar_url: str | None
    role: str
    joined_at: datetime


class GroupCreateResponse(BaseModel):
    id: str
    name: str
    description: str | None
    photo_url: str | None
    invite_code: str
    created_at: datetime


class GroupListItem(BaseModel):
    id: str
    name: str
    photo_url: str | None
    member_count: int
    members_preview: list[MemberAvatar] = []
    current_round: RoundSummary | None = None
    my_reading_progress: MyReadingProgress | None = None
    last_message_preview: LastMessagePreview | None = None
    last_activity_at: datetime | None = None


class GroupListResponse(BaseModel):
    groups: list[GroupListItem]


class GroupDetailResponse(BaseModel):
    id: str
    name: str
    description: str | None
    photo_url: str | None
    invite_code: str | None
    max_members: int
    member_count: int
    members: list[MemberSummary]
    current_user_id: str
    current_round: None = None
    created_at: datetime


class MessageResponse(BaseModel):
    message: str


class RegenerateCodeResponse(BaseModel):
    invite_code: str
    qr_url: str


class QrCodeResponse(BaseModel):
    qr_url: str


class MemberRoleUpdateRequest(BaseModel):
    role: Literal["admin", "member"]


class MemberRoleUpdateResponse(BaseModel):
    user_id: str
    role: Literal["admin", "member"]
    message: str

File: frontend/src/lib/types/group.ts
export type RoundStatus =
  | "nominating"
  | "voting"
  | "reading"
  | "reviewing"
  | "finished";

export interface RoundSummary {
  id: string;
  round_number: number;
  status: RoundStatus;
  book_title: string | null;
  book_author: string | null;
  book_cover_url: string | null;
  book_page_count: number | null;
}

export interface MyReadingProgress {
  current_page: number | null;
  total_pages: number | null;
  percentage: number;
}

export interface LastMessagePreview {
  sender_display_name: string | null;
  sender_avatar_url: string | null;
  content_text: string | null;
  content_type: string;
  created_at: string;
}

export interface MemberAvatar {
  user_id: string;
  display_name: string | null;
  avatar_url: string | null;
}

export interface GroupListItem {
  id: string;
  name: string;
  photo_url: string | null;
  member_count: number;
  members_preview: MemberAvatar[];
  current_round: RoundSummary | null;
  my_reading_progress: MyReadingProgress | null;
  last_message_preview: LastMessagePreview | null;
  last_activity_at: string | null;
}

export interface GroupListResponse {
  groups: GroupListItem[];
}

export interface GroupCreateResponse {
  id: string;
  name: string;
  description: string | null;
  photo_url: string | null;
  invite_code: string;
  created_at: string;
}

export interface MemberSummary {
  user_id: string;
  username: string | null;
  display_name: string | null;
  avatar_url: string | null;
  role: "admin" | "member";
  joined_at: string;
}

export interface GroupDetailResponse {
  id: string;
  name: string;
  description: string | null;
  photo_url: string | null;
  invite_code: string | null;
  max_members: number;
  member_count: number;
  members: MemberSummary[];
  current_user_id: string;
  current_round: null;
  created_at: string;
}

export interface MemberRoleUpdateResponse {
  user_id: string;
  role: "admin" | "member";
  message: string;
}
